mobile-browser-based-version/server/run_server.js

```diff
@@ -1,86 +1,293 @@
+
+  const receivedWeights = weightsDict[task][round]
+  if (Object.keys(receivedWeights).length < Math.ceil(peers.length * PEERS_THRESHOLD)) {
+    response.status(200).send({});
+    return;
   }
+
   // TODO: use proper average of model weights (can be copied from the frontend)
-  let serializedWeights = await averageWeights(Object.values(receivedWeights))
-  console.log(serializedWeights)
-  console.log(typeof serializedWeights)
-  let weights = msgpack.encode(Array.from(serializedWeights))
-  response.send({'weights': weights})
-  return
+  const serializedWeights = await averageWeights(Object.values(receivedWeights));
+  const weightsJson = JSON.stringify(serializedWeights);
+
+  if ((round - 1) % MODEL_SAVE_TIMESTEP == 0) {
+    const weightsPath = `${task}_${round}_weights.json`;
+    const milestonesPath = path.join(__dirname, 'milestones');
+    if (!fs.existsSync(milestonesPath)) {
+      fs.mkdirSync(milestonesPath);
+    }
+    fs.writeFile(path.join(milestonesPath, weightsPath), weightsJson, (err) => {
+      if (err) {
+        console.log(err);
+        console.log(`Failed to save weights to ${weightsPath}`);
+      } else {
+        console.log(`Weights saved to ${weightsPath}`);
+      }
+    });
+  }
+
+  let weights = msgpack.encode(Array.from(serializedWeights));
+  response.status(200).send({weights: weights});
+  return;
 }
 
-Promise.all(models.map((createModel) => createModel()))
+function sendDataInfo(request, response) {
+  const requestType = 'SEND_data_info';
 
+  if (!isValidRequest(request)) {
+    response.status(400).send(INVALID_REQUEST_FORMAT_MESSAGE);
+    console.log(`${requestType} failed`);
+    return;
+  }
+
+  const id = request.body.id;
+  const timestamp = request.body.timestamp;
+  const samples = request.body.samples;
+  const task = request.params.task;
+  const round = request.params.round;
+
+  if (!(task in dataDict)) {
+    dataDict[task] = {};
+  }
+  if (!(round in dataDict)) {
+    dataDict[task][round] = {};
+  }
+  dataDict[task][round][id] = samples;
+  response.status(200).send("Number of samples successfully received.");
+
+  logsAppend(request, requestType);
+  return;
+}
+
+function receiveDataInfo(request, response) {
+  const requestType = 'RECEIVE_data_info';
+
+  if (!isValidRequest(request)) {
+    response.status(400).send(INVALID_REQUEST_FORMAT_MESSAGE);
+    console.log(`${requestType} failed`);
+    return;
+  }
+
+  const body = request.body;
+  const id = body.id;
+  const timestamp = body.timestamp;
+  const task = request.params.task;
+  const round = request.params.round;
+
+  if (!(task in dataDict && round >= 0)) {
+    response.status(404).send(INVALID_REQUEST_KEYS_MESSAGE);
+    console.log(`${requestType} failed`);
+    return;
+  }
+
+  const latestDataDict = {};
+  // For each round...
+  for (let data of Object.values(dataDict[task])) {
+    // ... fetch the latest entry
+    for (let [id, samples] of Object.entries(data)) {
+      latestDataDict[id] = samples;
```

Comment:
this function here is a bit hard to read. distinguish meta data from training data or weights. maybe store each metadata element with a key in a dict or json? now it's hard for the reader to understand that this datadict contains mainly sample numbers, and not so easy to extend if it will contain all other metadata (rounds, delay information etc).

also a different thing, i didn't see where in the code the metadata and the weights get saved on the server and how frequently etc. can you make it more clear? (then users can more easily modify it or pick it up from there)